reanalysis/generate_pedigree.py

```diff
@@ -0,0 +1,140 @@
+"""
+generate a ped file on the fly using the sample-metadata api client
+additional (optional) argument will remove all family associations
+    - family structure removal enforces singleton structure during this MVP
+"""
+
+
+from typing import Dict, List
+from sample_metadata.apis import FamilyApi, ParticipantApi
+
+import click
+
+
+PED_KEYS = [
+    'family_id',
+    'individual_id',
+    'paternal_id',
+    'maternal_id',
+    'sex',
+    'affected',
+]
+
+
+def get_clean_pedigree(
+    pedigree_dicts: List[Dict[str, str]],
+    sample_to_cpg_dict: Dict[str, str],
+    singles: bool,
+) -> List[Dict[str, str]]:
+    """
+    swaps sample IDs for the internal CPG values
+    optionally strips out family structure and makes
+    all members into unrelated singletons
+    """
+
+    new_entries = []
+
+    for ped_entry in pedigree_dicts:
+
+        if ped_entry['individual_id'] not in sample_to_cpg_dict:
+            continue
+
+        # update the sample IDs
+        ped_entry['individual_id'] = sample_to_cpg_dict.get(ped_entry['individual_id'])
+
+        # remove parents and assign an individual sample ID
+        if singles:
+            ped_entry['paternal_id'] = ''
+            ped_entry['maternal_id'] = ''
+            ped_entry['family_id'] = ped_entry['individual_id']
+
+        else:
+            ped_entry['paternal_id'] = sample_to_cpg_dict.get(
+                ped_entry['paternal_id'], ''
+            )
+            ped_entry['maternal_id'] = sample_to_cpg_dict.get(
+                ped_entry['maternal_id'], ''
+            )
+
+        new_entries.append(ped_entry)
+
+    return new_entries
+
+
+def write_pedigree(clean_pedigree: List[Dict[str, str]], output: str):
+    """
+    take the pedigree data, and write out as a correctly formatted PED file
+
+    :param clean_pedigree: the PED content with sample names updated
+    :param output: file location to create
+    """
+    with open(output, 'w', encoding='utf-8') as handle:
+        handle.write('\t'.join(PED_KEYS) + '\n')
+        for entry in clean_pedigree:
+            handle.write('\t'.join([str(entry.get(key)) for key in PED_KEYS]) + '\n')
+
+
+def get_pedigree_for_project(project: str) -> List[Dict[str, str]]:
+    """
+    fetches the project pedigree from sample-metadata
+    list, one dict per participant
+    :param project:
+    :return: all content retrieved from API
+    """
+
+    return FamilyApi().get_pedigree(project=project)
+
+
+def ext_to_int_sample_map(project: str) -> Dict[str, str]:
```

Comment:
This endpoint doesn't return Dict[str, str], but List[Tuple[str, str]]. It does this because a participant can have multiple samples.

Reply:
very odd, I thought I had this wrapped in `dict()` to transform the results to a dictionary.... 

Hadn't occurred to me that the mapping here wouldn't be unique. Does that mean it's possible we end up with a one to many participant external ID to internal sample ID?

```
[(person1, sample1), (person1, sample2)]
```

Reply:
Yep, very plausible! Our maps for seqr already support many samples for one participant, and at the CPG a couple of projects have already encountered this.

Reply:
yikes, ok. This will need a bit of a re-think then. The pedigree generated here is intended for use in interpreting the results within a joint call... if multiple samples are present it will be tricky to know which sample to use 

parsing a VCF header to get a set of available samples and limiting the pedigree to those would work, but retrieving the same data from an MT is not so simple.

If this is integrated into the seqr pipeline using that same runtime PED file would be the logical choice, and this script can be discarded

Reply:
I haven't reviewed the seqr loading pipeline, @vladsaveliev, do we assume that a participant only has one sample?

Reply:
@illusional just having a think about this, it should be fine to duplicate the PED file row for every possible sample per participant. When the PED file is ingested for analysis, we'd just ignore all rows where the `individual ID` is not present in the callset. Not the cleanest solution, but Slivar already does something similar, and just logs the discrepant samples to STDERR and carries on with the present samples

Reply:
Yeah that sounds reasonable to me!

Reply:
Abandoning that idea... if a trio each have 2 samples that would have to expand out to 24 rows to ensure you have every possible sample combination. That's a recipe for chaos right there

Reply:
resurrecting this, with `itertools.product()` used to create every possible proband-mother-father sample combination obtained from the API. In realistic use cases I expect that the samples per participant in a single cohort will typically be 1, so the worst case scenario seems highly unlikely